### PR TITLE
Crafting snapshots [WIP] [CR]

### PR DESCRIPTION
**Pull request by mugling**
`Character::crafting_inventory` has two major limitations:
- It does not support an item `quality` consuming charges (eg. `WELD 2`)
- Vehicle mounted tools are provided by an ugly and buggy hack (`pseudo` items)

I'd like to propose a new `snapshot` class that uses the `visitable` interface to cache available items, charges and qualities. This PR outlines a potential design for review.

Three maps, `items`, `charges` and `quals` are used to cache all available sources which are stored via `item_location`. Queries against these caches are very efficient. Only when methods external to `snapshot` manipulate items is it necessary to call `refresh`. 

For crafting recipes with multiple requirements it is possible that an earlier selection removed an entry which also exists elsewhere in the caches. For example `screwdriver` will exist in both `items` and `quals` whereas `toolbox` provides multiple `quals`.

Rather than traversing each cache to remove all invalidated entries these are instead pushed to `snapshot::invalid`. Invalidated entries are then checked for on-demand. In theory two ordered sets can be handled in linear time but as `invalid` is likely to be at most a few entries `clean_invalid` takes the simpler approach.

Are there any more efficient designs without excessive complexity?

### CODE REVIEW

File: src/visitable.cpp [hunk 1]
```diff
@@ -297,8 +297,8 @@ static VisitResponse visit_internal( const std::function<VisitResponse( item *,
             return VisitResponse::ABORT;
 
         case VisitResponse::NEXT:
-            if( node->is_gun() || node->is_magazine() ) {
-                // content of guns and magazines are accessible only via their specific accessors
+            if( node->is_tool() || node->is_gun() || node->is_magazine() ) {
```
Comment (Coolthulhu) [inline src/visitable.cpp]: Which tools are affected?

Comment (mugling) [inline src/visitable.cpp]: None as of yet but it's a bug waiting to happen. The only contents for a tool should be a `magazine` which is definitely not available for crafting etc

**Comment by Coolthulhu** [comment]
> Only when methods external to snapshot manipulate items is it necessary to call refresh. 

Not sure if I understand correctly, but that sounds like you're trying to make the caches exist for more than just one craft attempt.

If so, `item_location` will need to gain the ability to unambiguously identify an item it is pointing at. That is, check if it still exists and check if it is the same item that was cached.

**Comment by mugling** [comment]
@Coolthulhu no - a calling `refresh` entirely recreates the caches using the same `position` and ` range`

**Comment by Coolthulhu** [comment]
Then it's simpler. Though it still must not crash if everything cached burns.

**Comment by mugling** [comment]
I don't follow?

**Comment by kevingranade** [comment]
What use cases is this trying to solve? crafting_inventory itself is an
intermediate data structure used for various things, so saying you're
extending it doesn't make it clear what the rationale and alternatives are.

In particular, the use cases surrounding caching available items from
within the crafting menu and caching items during crafting have different
requirements.

**Comment by Coolthulhu** [comment]
Could be useful for crafting from certain items, that's for sure.
`crafting_inventory` drops this data making it impossible to extend crafting requirements to more fine-grained, generic ones.

For example, it isn't possible to offer some rotten meat and some fresh meat as separate options, then consume only the fresh pieces so that the resulting product is fresh.

**Comment by kevingranade** [comment]
To extend on my previous comment, for tracking items while crafting is in
progress, I think a much simpler solution is to bundle the actual items
into a non-item object stored on the map. This resolves the
multiple-ownership problem that we encounter when both the map and a
crafting action separately reference items.
This would also remove the requirement on tracking items over time from the
crafting inventory class, allowing it to focus on efficiently caching data
needed by the crafting menu.

**Comment by mugling** [comment]
Closed for now, may reopen when resuming work on this

**Pull request closed without merging**